### DISCUSSION TITLE
BatteryService - Lessons learned

### DISCUSSION THREAD

**Discussion by helo-head** [HomeSpan/HomeSpan#156] [General]
It should have been obvious to me that the "BatteryService" is intended to only work as a secondary service to an accessory, not the primary service. The interesting part is that if you try to configure it as an accessory vs. a service, a tile will be shown and all of the characteristics are shown. In fact even a battery icon is shown once you open the tile. The only indications that something is wrong is the tile icon is a house and the statement "Not Supported" at the bottom of the tile. 

I was however disappointed that when configured correctly as a secondary service  no battery icon is shown. In fact the only way to determine the battery status is to select the tile, select settings and then scroll down to status. Thats a lot of work to determine the battery status ... Not sure when this changed in Homekit but its consistent between both the Home, Home+ and Eve applications. Eve does show a battery icon (its buried but its there) but I suspect this is an eve specific UI feature

**Comment by richonguzman** [comment]
hi, did you get to show as accessory?

**Reply by helo-head** [reply]
Have no tested since last year ... I now just use it as intended as secondary service. 

I would prefer to be able to show batteries as an accessory since I have a standalone whole house battery solution that I would like to be able to report on as an accessory vs a secondary service to some other bogus accessory like a light bulb.

**Comment by richonguzman** [comment]
I was using 3 separated ESP32 as a LED light and a Battery Indicator as accessory, so when battery is less than 30% it shows as "low battery" indicator in the home app

where did you add the battery indicator or , to which accessory did you add the battery_status?

**Reply by helo-head** [reply]
You can add it to any accessory where you think its relevant. In the example below my setup code added it to my garage door Vent/Fan

```
  new SpanAccessory();
    new DEV_Identify("M5 Vent","HomeSpan","M5Stack-002","M5 Fan","0.9");
    (new DEV_GarageFan())->setPrimary();
       new Characteristic::Name("M5 Fan"); 
        new DEV_Battery();
         new Characteristic::Name("M5 Battery");** 
```

Then in the dev_battery.h the constructor looks like 

```
struct DEV_Battery : Service::BatteryService { 

  SpanCharacteristic *level; 
  SpanCharacteristic *charge;
  SpanCharacteristic *low;
  
 DEV_Battery() : Service::BatteryService() {                   // constructor() method
    
 level=new Characteristic::BatteryLevel();                 // Percentage of total charge
 charge=new Characteristic::ChargingState();               //0-not charging, 1-charging, 2-Not chargeable
 low=new Characteristic::StatusLowBattery();               //0-normal, 1-low

  } // end constructor
```

**Reply by richonguzman** [reply]
so the way to add to any accessory is to declare it inside the accessory... will test as how it goes (the pin declaration how is it done?)
